OCPBUGS-17585: Tighten the rules for modifying Tuned Profiles

jmencak opened the pull request:
NTO operands allow updating Tuned Profiles.  This is intentional and by design
as some host information needs to be communicated back to the NTO operator,
but this also allows a successful local attacker potentially affect node-level
configuration of other cluster nodes.

This change addresses the situation in two ways.  First, scoped RBAC
permissions on Profile.status subresource is used to disallow Node-level
write access to Profile.spec.  Second, the Node resource is used to provide
status loops back to NTO using the kubelet credential to write an annotation
to the Node resource.

This change also simplifies the mechanism for accepting kernel
command-line parameters as calculated by the NTO operands.  Now, all NTO
operands must agree on the calculated kernel command-line parameters.

ClusterOperator/node-tuning now also reports operand version.  The
operand version changes only when all operand replicas have successfully
upgraded and are ready.  This information is used to block PPC
Reconcilation loop until the operator and operand RELEASE_VERSION agree.
This is a short-term measure to prevent from multiple node reboots
during upgrades.

Other fixes:
- Disallow MC updates during upgrades when kernel command-line
parameters of nodes within a MCP do not match.
- ClusterOperator/node-tuning object was sometimes giving false
information based on an old Metadata.Generation.

Resolves: OCPBUGS-17585

jmencak commented:
/payload-aggregate periodic-ci-openshift-release-master-ci-4.14-upgrade-from-stable-4.13-e2e-gcp-ovn-rt-upgrade 10

jmencak commented:
/test ci/prow/e2e-gcp-pao-updating-profile

jmencak commented:
/test e2e-gcp-pao-updating-profile

jmencak commented:
/retest

jmencak commented:
/cc @MarSik 
As discussed on Monday, only block PPC reconciliation loop on upgrades.  I've intentionally split this PR into 2 commits.  The first one was already reviewed by @dagrayvid .  Please take a look at mostly [the parts affecting PPC](https://github.com/openshift/cluster-node-tuning-operator/pull/775/commits/38b7d9437a924449292958821517adcc4969af4d).

MarSik commented on pkg/operator/mc.go:
```diff
@@ -259,6 +259,36 @@ func (pc *ProfileCalculator) getPoolsForNode(node *corev1.Node) ([]*mcfgv1.Machi
+		return nil, fmt.Errorf("invalid label selector %s in MachineConfigPool %s: %v", util.ObjectInfo(selector), pool.ObjectMeta.Name, err)
+	}
+
+	initialNodes, err := pc.listers.Nodes.List(selector)
```
I wonder if we should ignore nodes that are not ready or have scheduling disabled. Just to avoid a situation where a broken node (daemonset and/or kubelet) will prevent updates to the rest of the cluster.

jmencak replied:
I do not think I agree here, this would unnecessarily complicate the design and I"m assuming that at least during the upgrades, it is all hands on deck and people should deal with broken nodes.

MarSik replied:
But what if someone updates the PerfProfile while there is a node with stopped kubelet / nto daemonset? That Node object will have old bootcmdline data and will therefore prevent all the other nodes from updating to the new tuning.

jmencak replied:
Yes, but we're talking about a normal operation (not upgrades).  I agree such nodes need to be excluded and I'll adjust this PR to reflect it.

MarSik commented on pkg/performanceprofile/controller/performanceprofile_controller.go:
```diff
@@ -370,10 +372,24 @@ func validateUpdateEvent(e *event.UpdateEvent) bool {
+		return reconcile.Result{}, err
+	}
+
+	operatorReleaseVersion := os.Getenv("RELEASE_VERSION")
```
Is this maybe too strict? Well at least it allows changing the release version via the env if needed.

jmencak replied:
Not sure I know what you mean here, but I'm assuming this is just a comment without any action needed.

MarSik commented on pkg/tuned/controller.go:
```diff
@@ -918,6 +942,42 @@ func getNodeName() string {
+		return nil
+	}
+
+	_, err = c.kubeclient.CoreV1().Nodes().Update(context.TODO(), node, metav1.UpdateOptions{})
```
I think I need a better explanation about the security of this part. Is the annotation protected somehow? Is the kubelet client limited to updating only its own Node object?

In other words, please document the high level architecture and the rationale (the why) better here or in the commit message so we will not have to reread all the code to learn how it was supposed to work in the future.

jmencak replied:
Added a comment + the higher level architecture is already documented in the commit message.

jmencak commented:
/payload-aggregate periodic-ci-openshift-release-master-ci-4.14-upgrade-from-stable-4.13-e2e-gcp-ovn-rt-upgrade 10

jmencak commented:
/retest

jmencak commented:
@MarSik, all tests passed and the blocking job also succeeded successfully all 3 times I tried.  I belive I've hopefully also addressed your concerns.

yanirq commented on pkg/performanceprofile/controller/performanceprofile_controller.go:
```diff
@@ -370,10 +372,24 @@ func validateUpdateEvent(e *event.UpdateEvent) bool {
+
+	operatorReleaseVersion := os.Getenv("RELEASE_VERSION")
+	operandReleaseVersion := operatorv1helpers.FindOperandVersion(co.Status.Versions, tunedv1.TunedOperandName)
+	if operandReleaseVersion == nil || operatorReleaseVersion != operandReleaseVersion.Version {
```
Are there scenarios where we can end up with a long period of operand and operator version trying to match ?

jmencak replied:
Only during upgrades when we have failing nodes.  Once an upgrade is successful, operatorReleaseVersion and operandReleaseVersion will always match.

MarSik replied:
Also we were testing this with jmencak and MCO itself will not do changes to an MCP when there are failing nodes. We should document or link to a doc that says the proper way to handle a failing node is to remove it from the pool.

MarSik reviewed (approved):
/lgtm

I like it, just make sure the user will be visibly informed when the nodes do not agree and update is blocked. The same about the version mismatch, we should report that in the status (probably as a condition).

jmencak commented on pkg/operator/controller.go:
```diff
@@ -775,25 +765,30 @@ func (c *Controller) getProviderName(nodeName string) (string, error) {
 
-	bootcmdline := profile.Status.Bootcmdline
+	if ok := c.allNodesAgreeOnBootcmdline(nodes); !ok {
+		return fmt.Errorf("not all %d Nodes in MCP %v agree on bootcmdline: %s", len(nodes), pools[0].ObjectMeta.Name, bootcmdline)
```
Not sure what you mean. Tuned CR/object doesn't really use status.

jmencak commented:
> I like it, just make sure the user will be visibly informed when the nodes do not agree and update is blocked. The same about the version mismatch, we should report that in the status (probably as a condition).

Let's do this as a follow up PR, this needs to get in.

/label acknowledge-critical-fixes-only
